californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -508,11 +529,12 @@ private KeyToken createUnusedToken() {
 				KeyMID idByMID = new KeyMID(exchange.getCurrentRequest().getMID());
 				KeyToken idByToken = new KeyToken(exchange.getCurrentRequest().getToken());
 
-				exchangesByToken.remove(idByToken);
-
-				// in case an empty ACK was lost
-				exchangesByMID.remove(idByMID);
+				if (!exchange.getRequest().isMulticast() || isMulticastTimedOut(exchange)) {
+					exchangesByToken.remove(idByToken);
```

Comment:
Defined them in `NetworkConfig` and `NetworkConfigDefaults` in compliance with https://tools.ietf.org/html/rfc7390#section-2.5.